text/0000-binary-ops-specialization.md

```diff
@@ -45,88 +49,102 @@ where
-that satisfy the same parameterized `Borrow`. The rule for a crate defining
-such a type is to also define default impls of the new operator traits
-described in this RFC, where a generic type parameter bound by `Borrow`
-defines the operand type other than `Self`:
+fall into two categories. One category is different purpose-specific
+representations of an underlying data type that provides binary
+operators acting on itself. We'll call it the **base operand type** for the
+purposes of this proposal.
+For example, `String` is the standard owned counterpart of `str`,
+and `PathBuf` is this for `Path`. The base operand type itself is
+considered together with its operand type family for the following rule.
+Types in each such family usually implement `Borrow` to the base operand type,
+and their binary operator trait impls, as currently provided, tend to cover
+any possible pairs with the other types in the family.
+
+The rule for a crate defining such a type is to also define generic default
+implementations of the default operator overload traits described in
+this RFC, where a generic type parameter bound by `Borrow` to the base
+operand type defines the operand type other than `Self`:
 
 ```rust
-impl<Rhs> PartialEq2<Rhs> for String
+impl<Rhs> DefaultPartialEq<Rhs> for String
 where
     Rhs: ?Sized + Borrow<str>,
 {
-    default fn eq2(&self, other: &Rhs) -> bool {
+    default fn eq(&self, other: &Rhs) -> bool {
         &self[..] == other.borrow()
     }
 }
 ```
 
-The type parameter of the `Borrow` bound is the type that `Self`
-can also be borrowed as (which can be just `Self`). The role of `Borrow`
-therefore extends to restricting operand types of binary operators
-available for the implementing type.
+The type parameter of the `Borrow` bound is the base operand type for `Self`
+(which is `Self` in case the impl is defined for the base type itself).
+The new semantic of `Borrow` therefore extends the "acts the same"
+guarantee to binary operators available for the implementing type, which
+is already the case for the intra-type traits `Eq`, `Ord`, and `Hash`.
 Notably, the standard library largely maintains this stratification in the
-provided implementations of `Borrow` and Rust 1.0 operator traits;
+provided implementations of `Borrow` and the plain old operator traits;
 `PathBuf`/`Path` is a [problematic][issue55319] exception.
 
 Operator traits that take ownership of the operands are trickier to implement
-for non-`Copy` types: these should not work between two borrowed values
-to avoid allocations hidden in operator notation, while moving the
-owned operands into an operator expression may be non-ergonomic.
+for non-`Copy` types: these should not work between two borrowed
+values to avoid allocations or other side effects hidden in operator notation,
+while moving both owned operands into an operator expression
+may be non-ergonomic.
 A precedent is set in the `Add` implementation for `String` to only let
```

Comment:
More than a few are unhappy with this precedent, and would like to see it changed.

Reply:
I see no really good alternatives that could have been chosen with the non-specializable `Add` for Rust 1.0. There are probably other considerations for some owned types without extended `Borrow` families. What do linear algebra people do?